### ISSUE TITLE
Why not add support for ONVIF events for motion detection?

### ISSUE THREAD

**Issue by Aculeasis** [blakeblackshear/frigate#3485]
**Describe what you are trying to accomplish and why in non technical terms**
I want to use cameras' motion detectors via ONVIF events to decrease cpu usage.

**Describe the solution you'd like**
Add new motion detector type - something like the ONVIF motion events

**Additional context**
Many modern ip cameras have internal detectors (sound, move, zone etc.) and support ONVIF events. For example a connected my dahua cam to HA via ONVIF and I can use motions' binary sensors. I think it is a wonderful option for low hardware with coral tpu and a lot of cameras.
https://www.home-assistant.io/integrations/onvif/#supported-sensors

**Comment by NickM-27** [comment]
I'm not super familiar with ONVIF motion detection protocol, but frigate doesn't just use "if motion detected somewhere then do x", it is hyper focused on only detected motion based on how the config is setup and then also will ignore motion that occurs inside of motion masks (like a tree or something).

Frigate uses the location of motion to decide where to pull a region from to run object detection on it (frigate does not run object detection on the entire frame when motion occurs). So if ONVIF motion doesn't provide this info, it would be useless to frigate.

**Comment by NickM-27** [comment]
I looked through [the ONVIF protocol specifications](http://www.onvif.org/specs/core/ONVIF-Core-Specification.pdf) but it doesn't appear that this information is included in the spec in a way that is consumable by frigate

**Comment by Aculeasis** [comment]
> So if ONVIF motion doesn't provide this pixel based location of motion, it would be useless to frigate.

It is sad. Maybe hikivision protocol supports it, but it's proprietary.

**Comment by NickM-27** [comment]
I'm not sure if it needed to be closed, my second comment highlighted some options that could work depending on how it's implemented

**Comment by Aculeasis** [comment]
> The manually defined polygons / regions wouldn't realistically

This is it? 
![image](https://user-images.githubusercontent.com/42580940/179649967-182489b3-5a3f-4c60-a948-0afdfdf2d2d0.png)
I think this has only a region name.
![image](https://user-images.githubusercontent.com/42580940/179650320-442e2cb5-baf2-43bd-ad81-8298fd9ba54c.png)

Also, I tried to subscribe on events using [onvif_zeep](https://github.com/FalkTannhaeuser/python-onvif-zeep) for my inside camera, but all examples don't work and ONVIF device manager can't see the events. But HA works =\

**Comment by NickM-27** [comment]
I was referring to 

> Cell based motion detection determined by placing a grid over the video source and determining changes.

Not sure if you see anything like that

**Comment by Tyler-2** [comment]
Wouldn't this be useful in the sense that Frigate wouldn't need to do constant motion detection, and could just do motion detection on a buffered stream based on when ONVIF motion is reported?

**Comment by NickM-27** [comment]
> Wouldn't this be useful in the sense that Frigate wouldn't need to do constant motion detection, and could just do motion detection on a buffered stream based on when ONVIF motion is reported?

Frigates motion detection is not momentary, it keeps calculating an average background to know where motion is occurring as well as to know when to ignore motion because it's too large (like when a camera changes from color to IR).

**Comment by YBonline** [comment]
Can I request this be reopened? My server CPU ground to a halt today because I had a TON of motion activity (I believe in excess of 25 cameras, which is unusual), and this is a feature that other software like Shinobi and Blue Iris has. I understand its not a priority but I'm hoping it can be added at some point, I'm pretty confident I'm at my limit of adding more cameras without something to reduce CPU and memory usage substantially

**Comment by NickM-27** [comment]
we already have ideas for how to reduce motion cpu usage, and as has been explained above the motion data in onvif is not specific enough and also likely not fast enough to process data in real time like frigate does

**Comment by webgress** [comment]
Ok frigate is keeping track of motion areas. 
Why not start that from the ONVIF trigger minus few frames?
Effectively re-run the detection but only after the ONFIF signal.
It would add a bit of lag, that's true.
Is that not acceptable?

**Comment by NickM-27** [comment]
Frigate processes frames in real time, waiting on onvif data would back up the whole system, all so that motion detection can still be run in most cases anyway

**Comment by NickM-27** [comment]
for the reasons described above using the cameras motion detection will not be viable

**Comment by titovskiy** [comment]
@NickM-27 why not to user onvif motion sensor as helper? 

> Frigate uses the location of motion to decide where to pull a region from to run object detection on it (frigate does not run object detection on the entire frame when motion occurs). So if ONVIF motion doesn't provide this pixel based location of motion, it would be useless to frigate.

you describe that it not send motion position, but in some reasons with not so good camera quality it hurts cpu and other system resources

> , it keeps calculating an average background to know where motion is occurring as well as to know when to ignore motion because it's too large (like when a camera changes from color to IR).

but this not work as well, switching to IR trigger motion... 

so why not to give users ability to choose motion trigger?

**Comment by NickM-27** [comment]
Like has already been explained it doesn't provide enough information to work with Frigates real time object detection. And especially the new motion based features coming in 0.18.

**Comment by titovskiy** [comment]
> especially the new motion based features coming in 0.18

ooo!!! that's sounds good))

**Comment by titovskiy** [comment]
> especially the new motion based features coming in 0.18

ooo!!! that's sounds good))

**Comment by eldi0s** [comment]
One simple ask from the above is, to be able to save recordings based on the cameras integrated ONVIF motion detection API, and not use the integrated AI detection at all. Considering most up to date 30-50€ cameras have a reliable motion detection (zones, cars, persons, etc. configurable via app) this is a fair ask in reducing CPU utilization. Would enable people to use frigate on older hardware.

I have motioneye (no audio) with ffmpeg running with 4 cameras triggered via webhook API from HomeAssistant on a PI 3+. 
AgentDVR offers the same thing via ONVIF and including audio but has a paywall if you intend to use it remote.

Pre recording of 10-15 seconds prior the motion event would require permanent stream processing and recording nonetheless, but data can be discarded after couple of minutes (not a whole day of recording needed as implemented atm) if not saved permanently by a triggered event. Thats how motioneye is working.

**Comment by hawkeye217** [comment]
Most of the CPU usage in Frigate is due to object detection and video decoding. Frigate's motion detection algorithm is extremely lightweight, so even on older hardware it should not be a concern to configure motion retention for recordings and keep object detection disabled.

**Comment by titovskiy** [comment]
> Most of the CPU usage in Frigate is due to object detection and video decoding. Frigate's motion detection algorithm is extremely lightweight, so even on older hardware it should not be a concern to configure motion retention for recordings and keep object detection disabled.

Unfortunately no, even the now popular low cost Xeons are decently loaded with active movement in the frame, and this despite the fact that the detector goes through gpu (2070)

**Comment by hawkeye217** [comment]
You may be confusing object detection and motion detection. When you have object detection enabled, even with a configured GPU, Frigate has to use the CPU to prepare the inputs for inference. Disabling object detection should significantly reduce CPU usage. 

Motion detection uses very little CPU in comparison.

**Comment by NickM-27** [comment]
> One simple ask from the above is, to be able to save recordings based on the cameras integrated ONVIF motion detection API, and not use the integrated AI detection at all. Considering most up to date 30-50€ cameras have a reliable motion detection (zones, cars, persons, etc. configurable via app) this is a fair ask in reducing CPU utilization. Would enable people to use frigate on older hardware.

To be clear this is already possible and we are aware of users that do this via the create events API

**Comment by titovskiy** [comment]
> Frigate has to use the CPU to prepare the inputs for inference.

But this is unexpected, I thought that in case of motion detection, the entire frame is transmitted and then the GPU is already working on its definition... and then it turns out that the processor still performs a considerable task.. so apparently the same leaves, rain and other effects can load so much

And what is expected in the 18th version? Will there be another logic of transfer to processing?

**Comment by hawkeye217** [comment]
> I thought that in case of motion detection, the entire frame is transmitted and then the GPU is already working on its definition

No, see this section of the documentation: https://docs.frigate.video/configuration/masks/#further-clarification

**Comment by eldi0s** [comment]
> To be clear this is already possible and we are aware of users that do this via the create events API

Thanks

What about a retention period of minutes/hours instead of a whole day (I think this is the minimum atm) for buffering in RAM instead of SSD/MMC. That would significantly lower wear and tear on SSD/MMC drives.

**Comment by NickM-27** [comment]
> What about a retention period of minutes/hours instead of a whole day (I think this is the minimum atm) for buffering in RAM instead of SSD/MMC. That would significantly lower wear and tear on SSD/MMC drives.

You can set decimals so you can use whatever value you want. However you seem to misunderstand how frigate works. Any retention will lead to those relevant clips being written to the SSD from the RAM. But depending on your config only clips that are meant to be kept get moved to the SSD.

**Comment by eldi0s** [comment]
Tried the API and it's working great so far. All detections are tagged as onvif_motion and pre recording is working. But all events are treated as alarms not motion. I can live with that, but how can I deactivate the review function for alarms so that it's just recorded acc. retention policy for alarms?

**Comment by NickM-27** [comment]
In 0.18 you'll be able to mark it as a detection instead
